### PR TITLE
feat(limit): recreate order

### PR DESCRIPTION
**Pull request by alfetopito**
# Summary

Recreate LIMIT orders based on filled/cancelled/expired orders

- Open on its own modal
- Separated state (where needed)
  - independent settings
- Everything else identical to regular limit orders flow
- Suuuuuper not pretty, out of the scope

# To Test

1. Go to LIMIT orders history
2. Click on the dot menu
* Should have `recreate order` button 
image
3. Click on it
* Should load a "pretty" modal with the exact same data as used on the initial order
image
  - Same sell/buy tokens (should not be changeable)
  - Same sell/buy amounts
  - Order type (buy or sell order)
  - Same duration
  - Same recipient, if any
  - Same sub-type (fill-or-kill OR partially-fillable)
4. Place order
* Should take you back to the regular LIMIT orders screen, with the same values you had previously
* Settings state should remain the same as before

5. Repeat with different combinations of order type, sub-type, tokens, durations, etc
6. Placing regular LIMIT orders should remain the same as before
7. Does not affect SWAP or TWAP, so there shouldn't be any changes there either

### CODE REVIEW

**Comment by shoom3301** [comment]
One UX suggestion:
When you recreate an order from limit orders history tab, after the order creation would be great to automatically switch to open orders tab

**Review by shoom3301 — APPROVED** [review]
Very nice feature!

**Comment by alfetopito** [comment]
> One UX suggestion: When you recreate an order from limit orders history tab, after the order creation would be great to automatically switch to open orders tab

Turned into a big change, so it's in a new PR https://github.com/cowprotocol/cowswap/pull/3889

**Comment by elena-zh** [comment]
One more enhancement: in a mobile view, it would be great to scroll back to top when press on the recreate order button: 
image

**Comment by alfetopito** [comment]
> Hey @alfetopito , great job!
> 
> The most important issue I found (besides the navigation and UI :)) ), is that the button presents for TWAP orders (parent and parts). And when press on it, navigation buttons become missing on the TWAP form. Then, when go to the Limit orders page, you will see the modal to recreate a twap order. limits twaps twaps1
> 
> And WDYT about adding the button into the Order receipt modal? I think, it might be convenient to users to have it there.
> 
> Thanks!

Addressed the TWAP issue. Should no longer have the option there.
The second, I agree. Will do so in another PR.

As for the mobile move to top, I'll let that up to Michel when he gets to style this 😬

**Review by anxolin — COMMENTED** [review]
One small initial think I saw, not sure if related to this PR yet, is the price is not right (there's a small price difference, so each time I open the re-creation modal, I need to click on set market price)

I see the price is also not the old price (and shouldn't), so, which price is it using? 

image

File: apps/cowswap-frontend/src/common/state/alternativeOrder/atomFactories.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import { atom, Getter, PrimitiveAtom, SetStateAction, Setter, WritableAtom } from 'jotai'
+
+import { isAlternativeOrderModalVisibleAtom } from './atoms'
+
+export function alternativeOrderAtomGetterFactory<AtomValue>(
```
Comment (anxolin) [inline apps/cowswap-frontend/src/common/state/alternativeOrder/atomFactories.ts:5]: this is exported, but never used. Did you mean to export it?

**Comment by alfetopito** [comment]
> One small initial think I saw, not sure if related to this PR yet, is the price is not right (there's a small price difference, so each time I open the re-creation modal, I need to click on set market price)
> 
> I see the price is also not the old price (and shouldn't), so, which price is it using?
> image

You are right, there's something wrong with the price.
Will work on it

File: apps/cowswap-frontend/src/modules/limitOrders/hooks/useUpdateActiveRate.ts [hunk 2]
```diff
@@ -17,23 +17,24 @@ export interface UpdateRateCallback {
...
 
       updateRateState(update)
 
       const isSell = isSellOrder(orderKind)
 
       if (activeRate) {
         // Don't update amounts when rate is set from URL. See useSetupLimitOrderAmountsFromUrl()
-        if (currentIsRateFromUrl || isRateFromUrl) {
+        // Don't update amounts when rate is set from AlternativeOrder. See AlternativeLimitOrderUpdater
+        if (currentIsRateFromUrl || isRateFromUrl || currentIsAlternativeOrderRate || isAlternativeOrderRate) {
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/limitOrders/hooks/useUpdateActiveRate.ts:37]: Maybe this is related to the issue I posted? 
Shouldn't we update the rate for the alternative too?

File: apps/cowswap-frontend/src/modules/limitOrders/updaters/AlternativeLimitOrderUpdater.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
+import { useSetAtom } from 'jotai'
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/limitOrders/updaters/AlternativeLimitOrderUpdater.ts:1]: Do we really need to duplicate the updater? 
if we do in the future SWAP (maybe not super soon, but is possible), it feels we have way too many updaters, and I can see the updater explosion with alternative for all flows

**Review by anxolin — APPROVED** [review]
Cool! Soft approved. Im only a bit concern with the duplication of code and things getting even a bit more complex

Is it possible to reuse the smart component for the limit orders, so they really use even the same updaters but with the a flag saying they are the alternative order?

**Comment by elena-zh** [comment]
Hey @alfetopito , changes LGTM!
LMK please when I need to take a look into the price issue (or will this be addressed in a separate pr?)

**Comment by alfetopito** [comment]
> Hey @alfetopito , changes LGTM! LMK please when I need to take a look into the price issue (or will this be addressed in a separate pr?)

I'll do a separate PR since I already have other 2 PRs on top of this one

**Review by elena-zh — APPROVED** [review]
Approved then :)

File: apps/cowswap-frontend/src/common/state/alternativeOrder/atomFactories.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import { atom, Getter, PrimitiveAtom, SetStateAction, Setter, WritableAtom } from 'jotai'
```
[the remark this answers is not in the document]
Comment (alfetopito) [reply apps/cowswap-frontend/src/common/state/alternativeOrder/atomFactories.ts:1]: Makes sense! Will update.

File: apps/cowswap-frontend/src/modules/limitOrders/updaters/AlternativeLimitOrderUpdater.ts (hunk 3 above)
Comment (alfetopito) [reply apps/cowswap-frontend/src/modules/limitOrders/updaters/AlternativeLimitOrderUpdater.ts:1]: This one is what sets up the state, so it's not entirely a duplication.

**Review by shoom3301 — APPROVED** [review]

**Comment by alfetopito** [comment]
There's only 1 pending task related to mobile https://github.com/cowprotocol/cowswap/pull/3930#issuecomment-1970793960

I'll merge as is, since that can be addressed later.

**Pull request merged**